File: 백준/Silver/2840. 행운의 바퀴/행운의 바퀴.java
import java.io.*;
import java.util.Arrays;
import java.util.HashSet;
import java.util.StringTokenizer;
import java.util.Set;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        boolean possible = true;
        char[] answer = new char[N];
        int idx = 0;
        Set<Character> charSet = new HashSet<>();
        Arrays.fill(answer, '?');
        for (int i = 0; i < K; i++) {
            st = new StringTokenizer(br.readLine());
            int cnt = Integer.parseInt(st.nextToken());
            char alpha = st.nextToken().charAt(0);
            idx += cnt;
            if (answer[idx % N] == '?') {
                if (!charSet.contains(alpha)) {
                    answer[idx % N] = alpha;
                    charSet.add(alpha);
                } else {
                    possible = false;
                    break;
                }
            } else if (answer[idx % N] != alpha) {
                possible = false;
                break;
            }
        }

        if (!possible) {
            System.out.println('!');
        } else {
            String ans = "";
            for (int i = 0; i < N; i++) {
                ans += answer[(idx + (N-i)) % N];
            }
            System.out.println(ans);
        }
    }
}

File: 백준/Silver/2840. 행운의 바퀴/행운의 바퀴.py
N, K = map(int, input().split())
answer = ['?'] * N
idx = 0
possible = True
alpha_set = set()
for _ in range(K):
    cnt, alpha = input().split()
    idx = (idx + int(cnt)) % N
    if answer[idx] == '?':
        if alpha not in alpha_set:
            answer[idx] = alpha
            alpha_set.add(alpha)
        else:
            possible = False
            break
    elif answer[idx] != alpha:
        possible = False
        break

if not possible:
    print("!")
else:
    ans = ''
    for i in range(N):
        ans += answer[(idx+(N-i)) % N]
    print(ans)